### PR TITLE
feat(search): Updates child court ID retrieval to use Redis cache 

### PR DESCRIPTION
**Pull request by ERosendo**
This PR implements Redis caching for court data to optimize retrieval and improve performance .

Key changes:

- Adds `lib/courts.py` with the following components:

   - `MinimalCourtData`: dataclass to hold essential court attributes.
   - `get_minimal_list_of_courts`: Retrieves the complete list of court data as `MinimalCourtData` objects from the Redis cache or the database. If not found in the cache, it fetches from the database and caches the result for 24 hours.
   -  `get_active_court_from_cache`: Leverages Redis caching to fetch the complete list of courts and then filters them in-memory based on the `in_use` attribute.
   - `lookup_child_courts_cache`: Recursively finds all child court IDs for a given list of court IDs, utilizing the cached court list for efficient traversal.

- Updates the SearchForm class to use the `get_active_court_from_cache` method 

- Removes the previous `lookup_child_courts` helper method 

### Additional notes 

I identified a potential security and performance vulnerability while debugging of the `lookup_child_courts`(current method) implementation. The method currently uses all court IDs from the search filter to create cache keys, which helps avoid repetitive calculations for identical ID combinations. However, the logic does not validate if all provided IDs are valid court IDs. This allows an attacker to inject arbitrary characters into the court filter. This injection forces the application to:

1.  Bypass the cache, forcing the application to recalculate the court hierarchy, even when the same combination of valid court IDs has been processed before.
2.  Store a new, potentially redundant, entry in the `django_cache` table, which remains the for 30 days.

This vulnerability can lead to increased database load due to redundant calculations and an unnecessary growth in the cache table's size.

**Example:**

```
# Valid request, likely served from cache
- GET http://localhost:8000/api/rest/v4/search/?type=o&court=ca4%20ca5

# Malicious request, bypasses cache and adds a new cache entry.
- GET http://localhost:8000/api/rest/v4/search/?type=o&court=ca4%20ca5%20a

# Another malicious request, creating a different cache entry.
- GET http://localhost:8000/api/rest/v4/search/?type=o&court=ca4%20ca5%20ab
```

### CODE REVIEW

File: cl/lib/courts.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    Retrieves a list of active courts from the cached court data.
+
+    This method fetches the court data using `get_minimal_list_of_courts()`. it
+    then filters this list to include only courts where the 'in_use' attribute
+    is True.
+
+    :returns: A list of MinimalCourtData objects representing active courts.
+    """
+    courts = get_minimal_list_of_courts()
+    return [c for c in courts if c.in_use]
```
Comment (mlissner) [inline cl/lib/courts.py:71]: Why bother with this? Why not just filter to `in_use=True` in `get_minimal_list_of_courts`?

**Review by mlissner — APPROVED** [review]
This looks about right to me! A couple things:

1. I love all your comments, but I think you've gone too far in your docstrings, making them hard to maintain. This is just style, but generally you'd want docstrings to say the big picture, and leave out details like which fields are selected, and the steps the code takes. Certainly not a big deal, but I'm hunting for things to comment on at this point.

2. This code is going to impact every API and user query, so we need some tests of the edge cases and normal cases. I imagine you could mock the cache so that it returns MinimalCourtData objects without even needing to use faker. Hm. 

Overall looks great! I'm glad we're doing this and if you want to merge after adding the tests, that seems great to me.

File: cl/lib/courts.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def get_parent_ids_from_cache(courts: list[MinimalCourtData]) -> set[str]:
```
Comment (mlissner) [inline cl/lib/courts.py:74]: This doesn't actually involve the cache, so maybe:

```suggestion
def get_court_parent_ids(courts: list[MinimalCourtData]) -> set[str]:
```

**Comment by mlissner** [comment]
Oh, and I meant to return to the vuln. Do I gather that the changes here eliminate it or are there more changes we'd need in order to do so?

**Comment by ERosendo** [comment]
> Do I gather that the changes here eliminate it or are there more changes we'd need in order to do so?

Yep, this PR takes care of that vulnerability. We're now caching the court hierarchy differently, which means we've removed the dependency on user input(for cache keys).

File: cl/lib/courts.py (hunk 1 above)
Comment (ERosendo) [reply cl/lib/courts.py:71]: > Why not just filter to in_use=True in get_minimal_list_of_courts?

The reason I didn't just filter to `in_use=True` directly in the `get_minimal_list_of_courts` method is that we needed the full list of courts for another part of the system, specifically the method that looks up child court IDs.

If we'd limited `get_minimal_list_of_courts` to only the 'in_use' courts, we would've had to create a separate method or store another list to get all the courts anyway. By keeping the full list in `get_minimal_list_of_courts`, we can then filter it for the `searchForm` class, but still have the full set available for the child court ID lookup. It's really about avoiding duplicate lists

File: cl/lib/courts.py (hunk 2 above)
Comment (ERosendo) [reply cl/lib/courts.py:74]: yeah, makes sense. 👍

File: cl/lib/courts.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            if court.parent_court_id in courts_to_check
+        }
+
+        # Update the set of child courts found so far.
+        child_courts.update(new_child)
+        if new_child.isdisjoint(parent_court_ids):
+            # If no new children are parent courts, then stop iterating, as
+            # there are no more levels.
+            break
+        step = new_child
```
Comment (albertisfu) [inline cl/lib/courts.py:116]: I noticed that the courts, which point back to themselves as described in https://github.com/freelawproject/courtlistener/issues/4708#issuecomment-2769874600, can still cause an infinite loop here.

I did some testing, and I believe this approach could prevent the infinite loop in those cases:

```suggestion
        child_courts: set[str] = set()
    step: set[str] = set(court_ids)
    while True:
        new_child = {
            court.pk
            for court in courts_from_cache
            if court.parent_court_id in step
        }
        
        # Only consider IDs we haven't seen before to avoid cycles due to courts
        # pointing to themselves.
        new_ids = new_child - courts
        if not new_ids:
            break
        # Update the visited courts and set up the next iteration step
        courts.update(new_ids)
        step = new_ids
```

Also a test regarding this issue would be helpful to prevent regressions in the future.

**Review by albertisfu — CHANGES_REQUESTED** [review]
Thanks @ERosendo this looks pretty good. I just left a suggestion regarding courts that points back to themselves.

File: cl/lib/courts.py (hunk 3 above)
Comment (ERosendo) [reply cl/lib/courts.py:116]: Thanks @albertisfu! Sorry I forgot to add an update about those court. I checked with Bill about the self-referencing relationships, and he confirmed it's ok to remove the self-referencing relationship in our db.

File: cl/lib/courts.py [hunk 4]
```diff
@@ -91,30 +91,31 @@ def lookup_child_courts_cache(court_ids: list[str]) -> set[str]:
...
+            break
+
         # Update the set of child courts found so far.
-        child_courts.update(new_child)
-        if new_child.isdisjoint(parent_court_ids):
+        courts.update(new_ids)
+        if new_ids.isdisjoint(parent_court_ids):
             # If no new children are parent courts, then stop iterating, as
             # there are no more levels.
             break
```
Comment (albertisfu) [inline cl/lib/courts.py:118]: Is this still required? I think the previous break in `if not new_ids:` would stop the loop if no more childs to lookup?

Comment (albertisfu) [reply cl/lib/courts.py:118]: Great! This line could reduce the number of iterations in one. Let's keep it. Thanks.

**Review by albertisfu — APPROVED** [review]
